### PR TITLE
Chain namespace

### PR DESCRIPTION
**Pull request by Smoren**
Hi @markrogoyski

What do you think about new namespace with functionality of chain calls of another namespaces' methods?

For example:

```php
$input = [1, -1, 2, -2, 3, -3];

$iter = Chain::create($input)
    ->filterTrue(static function ($value) {
        return $value > 0;
    })
    ->pairwise();

$output = iterator_to_array($iter);
// [[1, 2], [2, 3]]
```

This is only a draft of PR. If you think this is a good idea I'm ready to implement all the methods and cover them by tests.

### CODE REVIEW

**Comment by Smoren** [comment]
Hi @markrogoyski 

I've added all the relevant methods, but they haven't been fully tested yet.

What do you think about this functionality?

**Comment by markrogoyski** [comment]
Hi @Smoren,

Sorry for the delay in responding. I don't have regular access to my PC and internet temporarily.

Thanks for the PR. I like the idea here.

A couple initial thoughts. I think a better name for this would be Stream. So like:

`Stream::of($data)->...`

Also, it probably makes sense to think about starting a stream from an infinite iterator, for example:

`Stream::count(1, 1)->...`

And similarly for random iteration as a starting point.

Anyways, cool stuff. Apologies that it may still be a few days before I can thoroughly review the PR and leave more detailed feedback.

Thanks,
Mark

**Comment by Smoren** [comment]
@markrogoyski

Thanks for the feedback!

I am very glad that you liked this idea.

Made suggested changes to the naming.

1. Namespace `Chain` renamed to `Stream`.
2. Method `Stream::create()` renamed to `Stream::of()`.

File: tests/Stream/SingleTest.php [hunk 1]
```diff
@@ -0,0 +1,1069 @@
...
+                [1, -1, 2, -2, 3, -3],
+            ],
+            [
+                [],
+                static function (iterable $iterable) {
+                    return Stream::of($iterable)
+                        ->filterTrue(static function ($value) {
+                            return $value > 0;
+                        });
+                },
```
Comment (markrogoyski) [inline tests/Stream/SingleTest.php]: Suggestion: For these tests, if you are passing an anonymous function, you can use simple [arrow function syntax](https://www.php.net/manual/en/functions.arrow.php) to simplify this expression.

You should be able to write it like this and get the same result you are going for:
```php
fn ($iter) => Stream::of($iter)
    ->filterTrue(fn ($x) => $x > 0),
```

**Comment by markrogoyski** [comment]
Hi @Smoren,

I suggest adding this function to the `Stream` class.

```php
    public function toArray(): array
    {
        $result = [];
        foreach ($this->iterable as $item) {
            $result[] = $item;
        }

        return $result;
    }
```

It would be a finalizer method on the stream, like this:
```php
$result = Stream::of([1, 2, 3, 4, 5, 6, 7, 8, 9])
    ->filterTrue(fn ($x) => $x % 2 === 0)
    ->pairwise()
    ->toArray();
```

I think this will simplify your unit tests, as you will no longer need to have a `foreach` loop in every unit test.

I have some other ideas for methods to add, but this PR is big enough, so we can add them in later on.

Mark

File: tests/Stream/SingleTest.php (hunk 1 above)
Comment (Smoren) [reply tests/Stream/SingleTest.php]: Done.

**Comment by Smoren** [comment]
Hi @markrogoyski

1. Tests: all anonymous functions replaced by arrow functions.
2. Method `toArray()` added and used in tests.
3. Method `toAsscociativeArray()` added (we need it using key-value iterators, not value only).

**Comment by Smoren** [comment]
Hi @markrogoyski

1. Tests: all the anonymous functions are replaced by arrow functions.
2. Method `toArray()` is added and used in tests.
3. Method `toAssociativeArray()` is added (we need it using key-value iterators, not value only) and used in tests.

**Comment by markrogoyski** [comment]
Hi @Smoren,

Thanks again for this great PR. I've merged it in. I'm going to continue developing the Streaming concept to add more features to it.

BTW, I removed `toAssociativeArray()` because it was not safe as written if `zip` is used, as the keys will themselves be an array, and trying to use an array as an associative array key will result in a TypeError. I will add a more flexible termination function to map custom keys via an anonymous function.

Mark

**Comment by Smoren** [comment]
Thank you Mark!

**Pull request merged**